skfda/misc/metrics/_sdtw_distances.py

```diff
@@ -0,0 +1,350 @@
+            " either two FDataGrid objects"
+        )
+
+    # if not (fdata1.n_samples == 1 and fdata2.n_samples == 1):
+    #     raise ValueError(
+    #         "Both fdata1 and fdata2 must contain a single sample"
+    #     )
+
+
+def _sdtw_divergence(
+    fdata1: FData,
+    fdata2: FData,
+    *,
+    gamma: 1.0,
+    cost: None,
+    check_cost: False,
+    eval_points: None
+) -> NDArrayFloat:
+
+    # For developers and PR:
+    # final format of fdata1 and fdata2 is required to be FDataGrid,
+    # sharing the same domain(1D) and codomain dimensions(1D...nD).
+    # No need for _check_compatible and _cast_to_grid:
+    # => would raise an error in the following situation:
+    # if fdata1.domain_range[0] != fdata1.grid_points[0].domain_range[0]
+    # (sdtw_divergence can deal with different domain_range and grid_points)
+
+    # check whether fdata1 and fdata2 are equivalent
+    _check_input_fdata(fdata1, fdata2)
+
+    if isinstance(fdata1, FDataBasis):
+        # check eval_points
+        # must be a tuple of one or two one-dim grid
+        # maybe an existing skfda function does the job ?
+        _check_input_evalpoints(eval_points)
+
+        if len(eval_points) == 1:
+            fdata1 = fdata1.to_grid(eval_points[0])
+            fdata2 = fdata2.to_grid(eval_points[0])
+
+        elif len(eval_points) == 2:
+            fdata1 = fdata1.to_grid(eval_points[0])
+            fdata2 = fdata2.to_grid(eval_points[1])
+
+    n1, n2 = fdata1.grid_points[0].size, fdata2.grid_points[0].size
+
+    if gamma <= 0:
+        raise ValueError(
+            "gamma was set to {} but must be positive".format(gamma)
+        )
+
+    # cost must satisfy:
+    # shape=(n1, n2),
+    # nonnegativity,
+    # indiscernability and symmetry
+    if cost is not None:
+        if isinstance(cost, List):
+            if not len(cost) == 3:
+                raise ValueError(
+                    "If the alignment cost matrices are pre-computed,"
+                    " then 'cost' must be a list of three two-dimensional"
+                    " numpy arrays"
+                )
+            else:
+                # shape of cost_12, cost_11 and cost_22
+                expected_shapes = [(n1, n2), (n1, n1), (n2, n2)]
+                for idx, c in enumerate(cost):
+                    if not (isinstance(c, NDArrayFloat) and c.ndim == 2):
+                        raise ValueError(
+                            "The elements of 'cost' must all be "
+                            "two-dimensional numpy arrays. "
+                            "The {}-th element does not satisfy the expected "
+                            "format".format(idx)
+                        )
+
+                    if not c.shape == (n1, n2):
+                        raise ValueError(
+                            "Cost matrix must have shape"
+                            " ({}, {})".format(n1, n2)
+                        )
+
+                    _check_shape_postive_cost_mat(
```

Comment:
I know it looks inappropriate here but otherwise, when the cost matrix does not satisfay these properties the output value of the algorithm does not make any sense. That's why I put a boolean `check_cost` (default: False) as input.

Should I let it here or is the user supposed to check it before ?

Reply:
I guess that if it is only done when `check_cost` is `True`, and it is false by default, it is not a problem.
By the way, the code can be made more compact (and better tested) by generating the costs if `cost` is a callable and then reusing the same code path as if `cost` where a sequence of `ndarray`.

Reply:
Ok, so only a `callable` as for the input `cost`.
Actually, I could raise a warning instead of a `ValueError`, that lets more flexibility to the user (but increases the risk of making wrong conclusion) but at least make the user aware of the that risk WDYT ?

Reply:
I didn't mean to allow only a callable. I wanted to say that after the callable is evaluated, the code path for a callable input is the same as for a `ndarray`, so both can be merged together.
I am not sure about the warning/exception thing. It is easier to handle an exception programatically, and you can always call the function again without checking if that is what you want. Also, if the algorithm require these preconditions maybe it shouldn't continue. Is there any case in which you would want to call the algorithm breaking the preconditions?

Reply:
Sorry I've been busy too.
So I followed your suggestion, the main code path is the following:
(1) pre-computations of cost matrices,
(2) Check the properties of each cost matrix if ``check_cost`` is True,
(3) Return the sdtw value